48_ima_incremental_attestation.md

```diff
@@ -0,0 +1,307 @@
+the differences in the IMA logs from the last (previous) attestation verification are
+transferred.
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+The goal of this enhancement is to reduce network bandwidth consumption as well as
+the verifier's CPU cycle consumption and increase overall scalability of
+Keylime when IMA attestation is being used.
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The verifier will need to maintain per-agent state that reflects the state
+of the IMA PCR from the last attestation along with the next-to-request
+IMA log entry. The verifier needs to be extended to request the
+IMA log starting at a specific entry in the IMA log. There are no changes
+required to the TPM quote.
+
+This proposal assumes that affinity of an agent with a verifier is
+maintained throughout the attestation. This allows a specific verifier
+to maintain the state of and agent in memory (IMA PCR and next-to-request
+IMA log entry). The Keylime verifier protocol seems to support this
+assumption since a verifier initiates the protocol for attestation by
+talking to an agent.
+
+To maintain backwards compatibility with older verifiers, the updated
+agent will return the entire IMA log when the verifier's request does
+not contain the number of the next IMA entry to return, which indicates
+and 'old' verifier.
+
+Similarly, to maintain backwards compatibility with older agents, an updated
+verifier has to be able to deal with the situation where the agent did
+not respond with the partial log but the entire IMA log. This situation
+is indicated by the agent not returning the number of the first IMA log
+entry it returned, thus the verifier will assume the log starts at
+the first entry.
+
+From the above it can be seen that the benefits of reduction of network
+bandwidth consumption and CPU cycle consumption are only going to take
+effect when updated verifiers and updated agents are being used.
+
+Upon restart of the verifier, it will request the IMA log from the
+first entry (entire log) when resuming communication with a particular
+agent. Therefore, startup of the IMA verification will be as costly
+(in terms of resource consumption) as it was before this enhancement.
+
+The per-agent state of the PCRs and the next-to-request IMA log entry
+will be maintained in the verifier's memory and be internally accessible
+using the unique ID of an agent (indexed by agent Id) as a
+search/selection criterion.
```

Comment:
Why maintain this in memory instead of in the DB? I don't think it's wrong, just curious about why that decision was made.

Reply:
It seems that a DB would be better in general.  Anything in memory is lost when the verifier reboots, and there may be need for analytics and forensics much later.

Reply:
> It seems that a DB would be better in general. Anything in memory is lost when the verifier reboots, and there may be need for analytics and forensics much later.

What does forensics do with temporary data holding the next-to-expect IMA measurement log entry line and the current state of the PCRs (10) related to IMA?

Reply:
For verifier restarts it's fine it will just need to ingest the whole log after a reboot. This might cause a storm of large attestation requests but I'm not sure it's a real problem since they'll be queued as the verifier gets to them.

If/when we move to a push rather than pull model that will be a bigger issue since that info will need to be serialized outside of the verifier since multiple verifiers might be working on the same pool of agents. But since that change is going to be a large architecture overhaul anyway I'm not sure we need to engineer it in now.

Reply:
> 
> 
> > It seems that a DB would be better in general. Anything in memory is lost when the verifier reboots, and there may be need for analytics and forensics much later.
> 
> What does forensics do with temporary data holding the next-to-expect IMA measurement log entry line and the current state of the PCRs (10) related to IMA?

My remark was that a persistent DB would be better **in general**.  There might be some individual fields that forensics does need.  OTOH, it's probably less code to store everything in the DB, and one never knows what might be needed after an attack.  

That specific state will also be needed if the verifier reboots.  You don't want the keylime verifier to say, "Sorry, I rebooted.  All you 10,000 attestors, please resend all 100,000 events now."